src/common/tests_dmg_helpers.c

```diff
@@ -0,0 +1,503 @@
+
+/* JSON output handling for dmg command */
+static int
+daos_dmg_json_pipe(const char *dmg_cmd, const char *dmg_config_file,
+		   char *args[], int argcount,
+		   struct json_object **json_out)
+{
+	char			*cmd_str = NULL;
+	char			*cmd_base = NULL;
+	struct	json_object	*obj = NULL;
+	int			parse_depth = JSON_TOKENER_DEFAULT_DEPTH;
+	json_tokener		*tok = NULL;
+	FILE			*fp = NULL;
+	int			pc_rc, rc = 0;
+
+	if (dmg_config_file == NULL)
+		D_ASPRINTF(cmd_base, "dmg -j -i %s ", dmg_cmd);
+	else
+		D_ASPRINTF(cmd_base, "dmg -j -o %s %s ",
+			   dmg_config_file, dmg_cmd);
+	if (cmd_base == NULL)
+		return -DER_NOMEM;
+	cmd_str = cmd_string(cmd_base, args, argcount);
+	D_FREE(cmd_base);
+	if (cmd_str == NULL)
+		return -DER_NOMEM;
+
+	D_DEBUG(DB_TEST, "running %s\n", cmd_str);
+	fp = popen(cmd_str, "r");
+	if (!fp) {
+		D_ERROR("failed to invoke %s\n", cmd_str);
+		D_GOTO(out, rc = -DER_IO);
+	}
+
+	/* If the caller doesn't care about output, don't bother parsing it. */
+	if (json_out == NULL)
+		goto out_pclose;
+
+	char	*jbuf = NULL, *temp;
+	size_t	size = 0;
+	size_t	total = 0;
+	size_t	n;
+
+	while (1) {
+		if (total + JSON_CHUNK_SIZE + 1 > size) {
+			size = total + JSON_CHUNK_SIZE + 1;
+
+			if (size >= JSON_MAX_INPUT) {
+				D_ERROR("JSON input too large\n");
+				D_GOTO(out_jbuf, rc = -DER_REC2BIG);
+			}
+
+			D_REALLOC(temp, jbuf, size);
+			if (temp == NULL)
+				D_GOTO(out_jbuf, rc = -DER_NOMEM);
+			jbuf = temp;
+		}
+
+		n = fread(jbuf + total, 1, JSON_CHUNK_SIZE, fp);
+		if (n == 0)
+			break;
+
+		total += n;
+	}
+
+	D_REALLOC(temp, jbuf, total + 1);
+	if (temp == NULL)
+		D_GOTO(out_jbuf, rc = -DER_NOMEM);
+	jbuf = temp;
+	jbuf[total] = '\0';
+
+	tok = json_tokener_new_ex(parse_depth);
+	if (tok == NULL)
+		D_GOTO(out_jbuf, rc = -DER_NOMEM);
+
+	obj = json_tokener_parse_ex(tok, jbuf, total);
+	if (obj == NULL) {
+		enum json_tokener_error jerr = json_tokener_get_error(tok);
+		int fail_off = json_tokener_get_parse_end(tok);
+		char *aterr = &jbuf[fail_off];
+
+		D_ERROR("failed to parse JSON at offset %d: %s %c\n",
+			fail_off, json_tokener_error_desc(jerr), aterr[0]);
+		D_GOTO(out_tokener, rc = -DER_INVAL);
+	}
+
+out_tokener:
+	json_tokener_free(tok);
+out_jbuf:
+	D_FREE(jbuf);
+out_pclose:
+	pc_rc = pclose(fp);
+	if (pc_rc != 0) {
+		D_ERROR("%s exited with %d\n", cmd_str, pc_rc % 0xFF);
+		if (rc == 0)
+			rc = -DER_MISC;
+	}
+out:
+	D_FREE(cmd_str);
+
+	if (obj != NULL) {
+		struct json_object *tmp;
+		int flags = JSON_C_TO_STRING_PRETTY | JSON_C_TO_STRING_SPACED;
+
+		D_DEBUG(DB_TEST, "parsed output:\n%s\n",
+			json_object_to_json_string_ext(obj, flags));
+
+		json_object_object_get_ex(obj, "error", &tmp);
+
+		if (tmp && !json_object_is_type(tmp, json_type_null)) {
+			const char *err_str;
+
+			err_str = json_object_get_string(tmp);
+			D_ERROR("dmg error: %s\n", err_str);
+			*json_out = json_object_get(tmp);
+
+			if (json_object_object_get_ex(obj, "status", &tmp)) {
```

Comment:
Is there a "status" obj in "error" obj? I only see a "Status" inside "response":

[boro-59:/home/huakuang/daos]$ dmg system list-pools -j -i -o ~/daos/install/etc/daos_control.yml
{
  "response": {
    "Status": 0,
    "Pools": null
  },
  "error": "failed to get MS response: client: code = 533 description = \"the server at boro-48:10001 refused the connection\""
}

Reply:
It's a new field added in this PR. It allows a DAOS status code to be relayed via RPC regardless of whether the call succeeds or fails.